charms/worker/k8s/src/charm.py

```diff
@@ -200,6 +200,40 @@ def _join_cluster(self):
             name = self.unit.name.replace("/", "-")
             self.api_manager.join_cluster(name, f"{str(address)}:{K8SD_PORT}", token)
 
+
+    @on_error(WaitingStatus("Waiting to remove nodes"), TypeError)
+    def _remove_nodes(self):
+        """Check for units that where removed, revoke their secrets and remove them from the k8s cluster."""
+        if relation := self.model.get_relation("cluster"):
+            units = [u.name for u in relation.units]
+            app_databag = relation.data.get(self.model.app, {})
+
+            units_to_remove = []
+            # Check if the app data contain tokens for units that do not
+            # exists anymore. If yes, clear token and remove node from cluster.
+            for unit_name in app_databag:
+                log.info(f"unit: {unit_name}")
+                still_exists = False
+                for u in units:
+                    if u == unit_name:
+                        still_exists = True
+                        break
+
+                if not still_exists:
+                    units_to_remove.append(unit_name)
+
+            if not units_to_remove:
+                return
+
+            status.add(ops.MaintenanceStatus("Removing nodes"))
+
+            for unit in units_to_remove:
+                del app_databag[self.unit.name]
+
+                node_name =unit.replace("/", "-")
+                # TODO: This will currently fail because worker removal does not work yet.
+                self.api_manager.remove_node(node_name)
+
```

Comment:
What if we leverage set operations instead?
```suggestion
    @on_error(WaitingStatus("Waiting to remove nodes"), TypeError)
    def _remove_nodes(self):
        """Check for units that where removed, revoke their secrets and remove them from the k8s cluster."""
        if relation := self.model.get_relation("cluster"):
            units = {u.name for u in relation.units}
            app_databag = relation.data.get(self.model.app, {})

            units_to_remove = set(app_databag.keys()) - units
            
            if not units_to_remove:
                return
                
            status.add(ops.MaintenanceStatus("Removing nodes"))

            for unit_name in units_to_remove:
                log.info(f" Unit: {unit_name} leaving cluster.")
                del app_databag[unit_name]

                node_name = unit_name.replace("/", "-")
                # TODO: This will currently fail because worker removal does not work yet.
                self.api_manager.remove_node(node_name)
```

Reply:
the [`ops.RelationDepartedEvent`](https://ops.readthedocs.io/en/latest/#ops.RelationDepartedEvent) will tell you which unit is leaving.  Would that help?

Reply:
That would help. But do we want that hybrid approach here? Some functions receiving the event, others don't...

Reply:
meh -- it's just there if you want it.  If the event happens to be the right one on the right relation -- it could be useful

Reply:
So, if we still need the set check I would just use that and not two different checks to archive the same thing. I understand that this would have a slightly better performance (we don't need to search through all units/secrets if we can facilitate the `RelationDeparted` event) but I would rather go for readability at this point. WDYT?